crates/rattler_networking/src/s3_middleware.rs

```diff
@@ -0,0 +1,434 @@
+    },
+}
+
+/// Wrapper around S3 client.
+#[derive(Clone, Debug)]
+pub struct S3 {
+    auth_storage: AuthenticationStorage,
+    config: HashMap<String, S3Config>,
+    expiration: std::time::Duration,
+}
+
+/// S3 middleware to authenticate requests.
+#[derive(Clone, Debug)]
+pub struct S3Middleware {
+    s3: S3,
+}
+
+impl S3Middleware {
+    /// Create a new S3 middleware.
+    pub fn new(config: HashMap<String, S3Config>, auth_storage: AuthenticationStorage) -> Self {
+        info!("Creating S3 middleware using {:?}", config);
+        Self {
+            s3: S3::new(config, auth_storage),
+        }
+    }
+}
+
+impl S3 {
+    /// Create a new S3 client wrapper.
+    pub fn new(config: HashMap<String, S3Config>, auth_storage: AuthenticationStorage) -> Self {
+        Self {
+            config,
+            auth_storage,
+            expiration: std::time::Duration::from_secs(300),
+        }
+    }
+
+    /// Create an S3 client.
+    ///
+    /// # Arguments
+    ///
+    /// * `url` - The S3 URL to obtain authentication information from the authentication storage.
+    ///     Only respected for custom (non-AWS-based) configuration.
+    pub async fn create_s3_client(&self, url: Url) -> Result<aws_sdk_s3::Client, Error> {
+        let bucket_name = url
+            .host_str()
+            .ok_or_else(|| anyhow::anyhow!("host should be present in S3 URL"))?;
+        if let S3Config::Custom {
+            endpoint_url,
+            region,
+            force_path_style,
+        } = self
+            .config
+            .get(bucket_name)
+            .unwrap_or(&S3Config::FromAWS)
+            .clone()
+        {
+            let auth = self.auth_storage.get_by_url(url)?;
+            let config_builder = match auth {
+                (
+                    _,
+                    Some(Authentication::S3Credentials {
+                        access_key_id,
+                        secret_access_key,
+                        session_token,
+                    }),
+                ) => {
+                    let sdk_config = aws_config::defaults(BehaviorVersion::latest()).load().await;
+                    aws_sdk_s3::config::Builder::from(&sdk_config)
+                        .endpoint_url(endpoint_url)
+                        .region(aws_sdk_s3::config::Region::new(region))
+                        .force_path_style(force_path_style)
+                        .credentials_provider(aws_sdk_s3::config::Credentials::new(
+                            access_key_id,
+                            secret_access_key,
+                            session_token,
+                            None,
+                            "pixi",
+                        ))
+                }
+                (_, Some(_)) => {
+                    return Err(anyhow::anyhow!("unsupported authentication method"));
+                }
+                (_, None) => {
+                    tracing::info!("No authentication found, assuming bucket is public");
```

Comment:
tbh i don't think the concept of public buckets makes much sense 🤔
for public buckets you could just use the http url of the bucket in your `channels` instead. for example, you could use `https://my-public-bucket.s3.eu-central-1.amazonaws.com/channel`.

i think we should just fail on no credentials found and advise people to use https for public buckets

Reply:
[`52add4e` (#1008)](https://github.com/conda/rattler/pull/1008/commits/52add4e02b370bb24a4604f68adc8534fda06b67)